core/io/json.cpp

```diff
@@ -55,7 +55,7 @@ static String _make_indent(const String &p_indent, int p_size) {
 	return indent_text;
 }
 
-String JSON::_print_var(const Variant &p_var, const String &p_indent, int p_cur_indent, bool p_sort_keys) {
+String JSON::_print_var(const Variant &p_var, const String &p_indent, int p_cur_indent, bool p_sort_keys, Set<const void *> &p_markers) {
```

Comment:
What's the point of having to pass in `p_markers`? Couldn't it be created by the class/function?

Reply:
`p_markers` can't be created by `_print_var()` because the function is recursive. It is created by the actual `print()` function which is the only caller of the recursive `_print_var()`.

`p_markers` is only useful during `print()`, so it does not make sense to make it a member of `JSON`. Making it a member variable would also introduce potential race condition if `JSON::print()` is called by multithreaded code.

`Variant::stringify()` requires passing a similar structure to track circular structure:
https://github.com/godotengine/godot/blob/364ea7f280a3f074795e542b16b1d0ec76cf6ce2/core/variant/variant.cpp#L1626